tests/php/ORM/DBReplicaTest.php

```diff
@@ -0,0 +1,230 @@
+        // Assert cms user logged in uses primary
+        $member = Member::get()->find('FirstName', 'cmsuser');
+        Security::setCurrentUser($member);
+        TestObject::get()->count();
+        $this->assertSame('primary', $this->getLastConnectionName());
+    }
+
+    public function testDataObjectMustUsePrimaryDb()
+    {
+        // Assert that DataList::getIterator() respect DataObject.must_use_primary_db
+        foreach (TestObject::get() as $object) {
+            $object->Title = 'test2';
+        }
+        $this->assertSame('replica_01', $this->getLastConnectionName());
+        foreach (Group::get() as $group) {
+            $group->Title = 'test2';
+        }
+        $this->assertSame('primary', $this->getLastConnectionName());
+        // Assert that DataQuery methods without params respect DataObject.must_use_primary_db
+        $methods = [
+            'count',
+            'exists',
+            'firstRow',
+            'lastRow'
+        ];
+        foreach ($methods as $method) {
+            (new DataQuery(TestObject::class))->$method();
+            $this->assertSame('replica_01', $this->getLastConnectionName(), "method is $method");
+            (new DataQuery(Group::class))->$method();
+            $this->assertSame('primary', $this->getLastConnectionName(), "method is $method");
+        }
+        // Assert that DataQuery methods with a param respect DataObject.must_use_primary_db
+        $methods = [
+            'max',
+            'min',
+            'avg',
+            'sum',
+            'column',
+        ];
+        foreach ($methods as $method) {
+            (new DataQuery(TestObject::class))->$method('ID');
+            $this->assertSame('replica_01', $this->getLastConnectionName(), "method is $method");
+            (new DataQuery(Group::class))->$method('ID');
+            $this->assertSame('primary', $this->getLastConnectionName(), "method is $method");
+        }
+    }
+
+    public function testRoutes()
+    {
+        // Create a custom rule to test our controller that should default to using a replica
+        $rules = Config::inst()->get(Director::class, 'rules');
+        $rules['test'] = TestController::class;
+        // Ensure that routes staring with '$' are at the bottom of the assoc array index and don't override
+        // our new 'test' route
+        uksort($rules, fn($a, $b) => str_starts_with($a, '$') ? 1 : (str_starts_with($b, '$') ? -1 : 0));
+        // Assert that normal routes use replica
+        $this->get('/test');
+        $this->assertSame('replica_01', $this->getLastConnectionName());
+        // Assert that Security route is forced to use primary
+        $this->get('/Security/login');
+        $this->assertSame('primary', $this->getLastConnectionName());
+        // Assert that dev route is forced to use primary
+        $this->get('/dev/tasks');
+        $this->assertSame('primary', $this->getLastConnectionName());
```

Comment:
We need to set `mustUsePrimary` back to `false` after each request, since it's a static property it's not really "per request" so much as it is "per PHP execution".

Reply:
Have changed to use dataProvider - mustUsePrimary is set back to false is setUp()